### ISSUE TITLE
Want a couple of exception design rules?

### ISSUE THREAD

**Issue by TravisCarden** [pepakriz/phpstan-exception-rules#151]
**tl;dr**: I have a couple of custom PHPStan rules related to exception design and handling that I want to contribute, and I'm wondering if you would be interested in them.

I maintain a library ([php-tuf/composer-stager](https://github.com/php-tuf/composer-stager)) with strict design rules. I've created a couple of custom PHPStan rules to help me follow them, and I'm interested in contributing them somewhere because I think they could be valuable to others, too. They're as polished as you would expect from something ad-hoc and project-specific, but I would be happy to generalize them and figure out how to write tests for them if you'd be interested in having them. Here they are. Please let me know if you would be open to a pull request.

- `Classes\MissingExceptionInterfaceRule` requires all exceptions to implement a common interface so that projects using my library can always fall back to catching it, knowing that it will capture _any_ of my exceptions.
- `Methods\ForbiddenThrowsRule` ensures that public methods (i.e., those that client code could call) never throw exceptions from third parties. This is the counterpart to the previous rule--it ensures that no third-party exceptions will leak through.

These rules, paired with those already in your package ensuring that all exceptions get annotated (`@throws`), can ensure nice, clean encapsulation for a library.

**Comment by mxr576** [comment]
@TravisCarden I like these ideas, I am also following these guidelines on my projects. As there is no activity on this issue maybe you could try to promote adding these new these new features directly to PHPStan or strict rules (not sure..). 

UPDATE: looks related  #148

**Comment by TravisCarden** [comment]
Ah, thanks, @mxr576. I didn't notice that the project had basically stagnated. I've created https://github.com/phpstan/phpstan/issues/8434 to propose this in PHPStan itself.

**Comment by TravisCarden** [comment]
@mxr576 since PHPStan itself wasn't interested in the rules, would you be interested in co-maintaining some kind of PHPStan DDD package in its own namespace? (I don't like to have things dependent solely on my user, in case I'm no longer able to (responsibly) maintain them at some point.)

**Comment by mxr576** [comment]
Thank for the offering. Unfortunately my outside of work capacity is already booked by other projects I have.

It just came into my mind, maybe this should not be handled by PHPStan? Maybe there is an existing solution for this like https://github.com/phparkitect/arkitect#available-rules?

**Comment by TravisCarden** [comment]
Most understandable, @mxr576. Thanks for sharing https://github.com/phparkitect/arkitect. I haven't seen that before.
